### ISSUE TITLE
BUG: Stuck at 35% Loading

### ISSUE THREAD

**Issue by Mezalion** [DarkflameUniverse/DarkflameServer#277]
### Make sure you've done the following:

- [X] I have checked that this issue has not already been reported.

- [X] I have validated that this issue is not a syntax error of either MySQL or SQLite.

- [X] I have pulled the latest version of the main branch of DarkflameServer and have confirmed that the issue exists there.

### Issue Description

Probably linked to #106 and #55

Multiple People reported this now: When trying to connect to a Server they get stuck at 35%. Different people on Different Server. When using their Phone as a HotSpot they could connect however, and when trying to connect to a different Server with the same Client everything works. After some experimenting we found out that the Server drops the Datatransfer-Speed to around 20kb/s at some point in the Loading. Not sure wether its a DLU Bug or Server-Setup Problem. 
When cloning the Server to a new Hoster it works however. 
Some Client Logs and Server Logs below.
[log_legouniverse_12-20-21_14-16-25__10420.log](https://github.com/DarkflameUniverse/DarkflameServer/files/7745602/log_legouniverse_12-20-21_14-16-25__10420.log)
`Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:47] [Server] Unknown packet ID from master: 21
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:51] [MasterServer] Got affirmation request of transfer 13
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:52] [AuthPackets] Received client version: 171022
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:52] [WorldServer] User test authenticated with correct key.
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:52] [Character] Loaded xmlData for character BrightSmartyAlgae (426)!
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:52] [User] Loaded 426 as it is the last used char
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:55] [WorldServer] Received level load complete from user.
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:55] [WorldPackets] Sent CreateCharacter for ID: 1152921508901814698
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:12:55] [ControllablePhysicsComponent] Using patch to load minifig physics
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:16:24] [TotalTimePlayed] Time since last save: 209
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:16:24] [Character] Saved character to Database in: 0.003584s
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:16:24] [WorldServer] Deleting player 1152921508901814698
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:16:24] [Player] Deleted player
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:16:24] [TotalTimePlayed] Time since last save: 0
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:16:24] [Character] Saved character to Database in: 0.000840s
Dec 20 13:16:31 Memezalions-Schachtel MasterServer[88776]: [20-12-21 13:16:24] [UserManager] Deleted user 30`
 

### Reproduction steps

Connecting to a Server using Home Network.
Creating a Char works.
Launching it does not, stuck at Loading into the Cutscene/when crashing the Client and restarting it does not load into VE and stops Loading at 35%. After some Time it sometimes loads to 55%, but not further. 

### Expected Behavior

To Launch the Cutscene/Load into VE

### Environment

Allways Ubuntu, sometimes in a VM, sometimes G-Cloud, Hetzner, Contabo... Appears all over the Board. Some people fixed it with Disabling their VPN if they used one.

**Comment by monst3rmash** [comment]
Sometimes after being stuck at the loading screen at 35%, I get kicked back to login screen with the error Message "Lost connection to world server"

[log_legouniverse_12-20-21_14-36-42__13628.log](https://github.com/DarkflameUniverse/DarkflameServer/files/7745664/log_legouniverse_12-20-21_14-36-42__13628.log)

**Comment by MichaelFedora** [comment]
This has something to do with being on a weird internet provider, particularly the smaller ones. As far as I understand it, the smaller network is something of a subnet, which causes this server (whether because of raknet or otherwise) to freak out and drop the connection. My brother has this issue with the same client I do, same server (remote), but he is under one of those smaller internet providers.

I am wondering if creating another interface with lcdr's [raknet shim mod](https://github.com/lcdr/raknet_shim_dll) in addition to raknet would work? Not sure.

**Comment by Mezalion** [comment]
> This has something to do with being on a weird internet provider, particularly the smaller ones. As far as I understand it, the smaller network is something of a subnet, which causes this server (whether because of raknet or otherwise) to freak out and drop the connection. My brother has this issue with the same client I do, same server (remote), but he is under one of those smaller internet providers.
> 
> Somehow is was able to load into VE and once connected he was fine, but never since then. So I believe this bug is simply when connecting/loading from the World Server, but once the handshake is done everything seems to work properly.
> 
> I am wondering if creating another interface with lcdr's [raknet shim mod](https://github.com/lcdr/raknet_shim_dll) in addition to vanilla raknet would work? Not sure.

Was my first thought as well (what you are referring to is named CGNAT afaik), but if it would be CGNAT we wouldnt be able to connect in the first place. And i think the Chars are transfered from the World Server, so they wouldnt load as well if it were only the World-Server.

**Comment by MichaelFedora** [comment]
The client log leaves much to be desired:


```log
20:11:45	2	GUI	Sending login request...
20:11:45	2		RakNet base connection to Auth acquired.  Sending version...
20:11:46	2		Connection to Auth confirmed. (NetID 0)
20:11:46	2		Submitting Login info to Auth...
20:11:46	2		Stamp PASSPORT_AUTH_LEGOINT_RECEIVED(44) at 14000 (start+0, last+0)
20:11:46	2		Attempting to connect to World...
20:11:46	2		Auth rejected/terminated connection.
20:11:47	2		RakNet base connection to World acquired.  Sending version...
20:11:47	2		Connection to World confirmed. (NetID 1)
20:11:47	2		Submitting Validation to World... (accnt: #####, key: #####)
20:11:47	2		Received Character List
...
20:11:56	2		Sending Character Login Request...
...
20:11:56	2		Received TransferToWorld (ip #.#.#.#, port 3004)
20:11:56	2		Attempting to connect to World...
20:11:56	2		Connection to 'Unknown Server' Dropped. (ID 1, IP #.#.#.#, Port 3001, Reason: 2)
20:11:57	2		RakNet base connection to World acquired.  Sending version...
20:11:57	2		Connection to World confirmed. (NetID 2)
20:11:57	2		Submitting Validation to World... (accnt: #####, key: #####)
20:11:57	2		Not requesting character list... (accnt: #####, key: #####)
20:11:57	2		Received Load Zone, ID (1000:1:0)
...
20:12:01	2		Received Create Character, ID ##### - Name #####
...
20:17:30	2		Connection to World Dropped. (ID 2, IP #.#.#.#, Port 3004, Reason: 4)
```

The server log also had nothing relevant besides creating the character. I could see his character in VE while he was stuck loading, but it would never finish for him.

**Comment by Mezalion** [comment]
Server-Logs said he is logging of the moment the Connection slows down if i see that right. Although he definetly did not.

**Comment by AxtonF** [comment]
Is there anything new to this? I have the same problem and could provide data or something if that is needed. Does anybody know of a fix. I already tried many different things like another OS and using a VPN. I also have a strong hunch this could be due to the router. I have a cable connection from Vodafone in NRW Germany and have already had problems with VPNs and other stuff with the router it's the Technicolor CGA6444VF.

**Comment by Mezalion** [comment]
Someone suggested that trying to TP the Person while they are loading in. It seems to work in some Edge-Cases where the Loading-Bug is client-side. But aside from that we found out that migrating the Server to another Hoster seemed to have fixed the Issue for some, but enabled it for others, even tho the Firewalls are allmost the same.

**Comment by averysumner** [comment]
Has this issue persisted?

**Comment by monst3rmash** [comment]
Sadly yes, still can’t play. Workaround for me was using a vpn but I don’t have access to one anymore.

**Comment by Mezalion** [comment]
Yep, still persistent. Someone on the LUCH (think it was either MM or Mega) said its a Client-Side-Thing, some other people said its something with RAK-Net.

**Comment by monst3rmash** [comment]
I tried it with different clients and computers in the same network and it don’t work. But with the same Notebook at my friends house with his wifi it works, so I don’t think it’s a client side bug.

**Comment by AxtonF** [comment]
> I tried it with different clients and computers in the same network and it don’t work. But with the same Notebook at my friends house with his wifi it works, so I don’t think it’s a client side bug.

It can still be a client-side bug though. This issue only affects certain network configurations (otherwise it wouldn't be solvable with a VPN), that's why it works at your friend's place. But the question is if this misbehavior under certain networks is a problem with the server code or the client code.

**Comment by averysumner** [comment]
I'm thinking that this is an issue with the WorldServer being unreachable on your network for some reason.

**Comment by AxtonF** [comment]
> I'm thinking that this is an issue with the WorldServer being unreachable on your network for some reason.

Why though? The only difference to the auth server is the port right? Why should a change in port make the server unreachable? And at least for me it sometimes also stopped at 55% or sometimes it even loaded me into the world perfectly. If the server became unreachable why only sometimes?

**Comment by EmosewaMC** [comment]
Does this issue still persist?  Please try rebuilding a server from scratch to start from a fresh known state and ensure you are using an unpacked client to setup the server for the easiest setup.

**Comment by EmosewaMC** [comment]
Checking in with others on the issue.  Thank you for the detailed posts and I'll get back to you on possible solutions

**Comment by EmosewaMC** [comment]
So something I observed is around 1400 bytes in length (1492 - 28) RakNet splits packets up and then puts them back together when the client receives them.  Its possible there could be something odd going on with the ReliabilityLayer if you wanted to look there for the issue, but ive got no more leads besides that for now

**Comment by EmosewaMC** [comment]
Ive found something that may address this issue.  Will push a fix and tag when the possible fix is in.  Basically we are going to reduce the MTU size by ~300 and see if smaller packets allow them to be delivered.

**Comment by EmosewaMC** [comment]
A fix has been merged to main.  Can you update and confirm is the issue persists?

**Comment by Mezalion** [comment]
If noone does it prior, ill do it around End of January when i have time again. Need to rebuild everything cause i had to delete my VPS, hence the delay.

**Comment by GoldenDewMelon** [comment]
I'm having this issue as well. My ISP is Spectrum. Hotspot works. Doesn't work on my network with or without VPN.
[log_legouniverse_03-25-23_00-27-20__44072.log](https://github.com/DarkflameUniverse/DarkflameServer/files/11068300/log_legouniverse_03-25-23_00-27-20__44072.log)

**Comment by EmosewaMC** [comment]
This is an ISP issue and has been proven over time to be unrelated to DLU networking.  Moving to Low priority due to this.

**Comment by TecCheck** [comment]
Any news on this? I tried to find an answer to why the UDP traffic has problems but couldn't find any answers at all. 

I looked into it using Wireshark. My computer receives all packets normally until it connects to the world server. The world server then sends a lot of packets (some of which bigger than 1000 bytes) and apart from a few early packets none of them get through and I only get packets with a size less than 30 bytes

**Comment by EmosewaMC** [comment]
This is similar to 55%.  Something related to tcp/udp having something change which causes the issue.  It's a won't fix on our end and the workaround is to switch network configuration (Hotspot, different wifi etc.).  Most users do no experience the issue altogether and replacing our networking interface at the moment is a tall order.

**Comment by aronwk-aaron** [comment]
Closed as won't fix
The only other work around known would be to integrate udp<->tcp into DLU.
